java/compaction/compaction-core/src/main/java/sleeper/compaction/strategy/CompactionStrategyIndex.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2022-2024 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.compaction.strategy;
+
+import sleeper.core.partition.Partition;
+import sleeper.core.statestore.FileReference;
+import sleeper.core.table.TableStatus;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class CompactionStrategyIndex {
+    private final List<FilesInPartition> filesInLeafPartitions;
+
+    public CompactionStrategyIndex(TableStatus tableStatus, List<FileReference> allFileReferences, List<Partition> allPartitions) {
+        Map<String, FilesInAscendingOrder> filesWithNoJobIdByPartition = new HashMap<>();
+        Map<String, List<FileReference>> filesWithJobIdByPartition = new HashMap<>();
+        Set<String> leafPartitionIds = allPartitions.stream()
+                .filter(Partition::isLeafPartition)
+                .map(Partition::getId)
+                .collect(Collectors.toSet());
+
+        allFileReferences.stream()
+                .filter(file -> leafPartitionIds.contains(file.getPartitionId()))
+                .forEach(file -> {
+                    String partitionId = file.getPartitionId();
+                    if (file.getJobId() == null) {
+                        filesWithNoJobIdByPartition.computeIfAbsent(partitionId, key -> new FilesInAscendingOrder()).add(file);
+                    } else {
+                        filesWithJobIdByPartition.computeIfAbsent(partitionId, key -> new ArrayList<>()).add(file);
+                    }
+                });
+        this.filesInLeafPartitions = leafPartitionIds.stream()
+                .map(partition -> new FilesInPartition(tableStatus, partition,
+                        filesWithNoJobIdByPartition.getOrDefault(partition, new FilesInAscendingOrder()).values(),
+                        filesWithJobIdByPartition.getOrDefault(partition, List.of())))
```

Comment:
It seems like if there are no files for the partition we can just ignore that partition?